File: algorithms/D/Detonate the Maximum Bombs/Detonate the Maximum Bombs.cpp
class Solution {
public:
    double eucDis(int x1,int y1,int x2,int y2)
    {
        double temp=pow(x1-x2,2)+pow(y1-y2,2);
        return sqrt(temp);
    }
    void dfs(int node,vector<int>&vis,vector<int>graph[],int &c)
    {
        c++;
        vis[node]=1;
        for(auto i:graph[node])
        {
            if(!vis[i])
            {
                dfs(i,vis,graph,c);
            }
        }
    }
    int maximumDetonation(vector<vector<int>>& bombs) {
        
        int n=bombs.size();
        vector<int>graph[n+1];
        int i,j;
        for(i=0;i<n;i++)
        {
            bombs[i].push_back(i);
        }
        for(i=0;i<n;i++)
        {
            
            for(j=0;j<n;j++)
            {
                if(i==j)
                    continue;
                double dis1=eucDis(bombs[i][0],bombs[i][1],bombs[j][0],
                               bombs[j][1]);
                
                if(dis1<=bombs[i][2])
                {
                    int node1=bombs[i][3];
                    int node2=bombs[j][3];
                    graph[node1].push_back(node2);
                }
            }
        }
       
        int maxi=1;
        for(i=0;i<n;i++)
        {
            int c=0;
            if(graph[i].empty())
                continue;
            vector<int>vis(n+1,0);
            
            dfs(i,vis,graph,c);   
            maxi=max(maxi,c);
            
            
        }
        return maxi;
    }
};

File: scripts/algorithms/D/Detonate the Maximum Bombs/Detonate the Maximum Bombs.java
// Runtime: 234 ms (Top 36.67%) | Memory: 42.3 MB (Top 94.92%)
class Solution {
    /*
    Make directed graph
    u -> v means, v is in the range of u
    check from which node maximum nodes can be reached and return the number of nodes reached
    */
    public int maximumDetonation(int[][] bombs) {
        Map<Integer, List<Integer>> graph = new HashMap<>();

        int n = bombs.length;
        for(int i = 0; i< n; i++){
            graph.put(i, new ArrayList<>());
            for(int j = 0; j< n; j++){
                if(i == j) continue;
                if(inRange(bombs[i], bombs[j]))
                    graph.get(i).add(j);
            }
        }

        int max = 0;
        for(int i = 0; i< n; i++){
            max = Math.max(max, dfs(i, graph, new HashSet<>()));
        }
        return max;
    }

    private boolean inRange(int[] u, int[] v){
        // (x-a)^2 + (y-b)^2 = R^2 -> point (a, b) is at border
        // (x-a)^2 + (y-b)^2 < R^2 -> point (a, b) is inside the circle
        // (x-a)^2 + (y-b)^2 > R^2 -> point (a, b) is outside the circle
        return Math.pow(u[0]-v[0], 2) + Math.pow(u[1]-v[1], 2) <= Math.pow(u[2], 2);
    }

    private int dfs(int node, Map<Integer, List<Integer>> graph, Set<Integer> visited){
        if(visited.contains(node)) return 0;
        visited.add(node);
        int res = 0;
        for(int neigh: graph.get(node)){
            res += dfs(neigh, graph, visited);
        }
        return res + 1;
    }
}